caddyconfig/caddyfile/adapter.go

```diff
@@ -109,6 +109,18 @@ type Unmarshaler interface {
 	UnmarshalCaddyfile(d *Dispenser) error
 }
 
+// EmbeddedMatcherUnmarshaler is the same as Unmarshaler, but will merge with any existing matchers of the the same type
+// from already parsed named matchers. The named matchers in embeddedMatchers should be merged.
+type EmbeddedMatcherUnmarshaler interface {
+	UnmarshalCaddyfileWithEmbeddedMatchers(d *Dispenser, matchers map[string]caddy.ModuleMap, embeddedMatchers []string) error
+}
+
+// BlockUnmarshaler is the same as Unmarshaler, but will compile named matchers to be passed to the next directive.
+// These are matcher operations like `not`.
+type BlockUnmarshaler interface {
+	UnmarshalCaddyfileWithMatchers(d *Dispenser, matchers map[string]caddy.ModuleMap) error
+}
```

Comment:
I'm not sure I like adding _two_ new interfaces here. Can we combine these into one? I figure the `BlockUnmarshaler` usecase would be covered by using a `EmbeddedMatcherUnmarshaler` with an empty `embeddedMatchers []string`, no?

Reply:
The problem is detection: when do you know to use an empty `embeddedMatchers`?

For example, in the following scenario:

```
@matcher1 {
    path /app
}

@matcher2 {
    path /other
    header Foo bar
}

@matcher3 {
    @matcher1
    not @matcher2
}
```

The intended outcome should be equivalent to:
```
@matcher3 {
    path /app
    not {
        path /other
        header Foo bar
    }
}
```
You need to know not to pass the `embeddedMatchers` to the `not` block, otherwise the above will get parsed as:
```
@matcher3 {
    path /app
    not {
        path /app /other
        header Foo bar
    }
}
```

Reply:
It could be solved by always passing in `embeddedMatcher` and then the implementation of `not` (and eventually `any` and `all`) to ignore it.

Reply:
Yeah I think we can pass `embeddedMatchers` to `not` anyways, and it can ignore it. I think it's probably clear that `not` is sorta like a scope, so it shouldn't inherit anything outside of it anyways. `any` and `all` would probably do the same as well.

Reply:
Cool. This is done and ready for review.